code/datums/music/music.dm

```diff
@@ -0,0 +1,263 @@
+/datum/music
+	/// The file played
+	var/sound_file
+	/// Whether the file loops - If false, the music automatically is removed once it reaches the end
+	var/does_loop = FALSE
+
+	/// The base volume to play at, not accounting for fading
+	var/base_volume = 100
+
+	/// Higher priority music plays over lower priority ones. If a looping music is masked by a higher priority music, it will start playing once the higher priority ends if it hasn't been removed.
+	var/priority = 0
+
+	var/client/target
+
+	var/fade_volume = 1
+	var/fade_volume_target = 1
+	var/fade_rate = 0 // per-second
+	var/sound/sound_datum
+	var/is_fading = FALSE
+
+/datum/music/New(client/C, _fade_volume = 1)
+	..()
+	C.active_music += src
+	target = C
+	fade_volume = _fade_volume
+	START_PROCESSING(SSprocessing, src)
+	if((target.prefs.toggles & PREFTOGGLE_SOUND_AMBIENCE) && (!target.playing_music || target.playing_music.priority < priority || (target.playing_music.priority == priority && target.playing_music.does_loop && !does_loop)))
+		unmask()
+	else if(!does_loop)
+		qdel(src)
+
+/datum/music/Destroy()
+	. = ..()
+	does_loop = TRUE // hacky code to stop runtimes from qdel loops
+	mask()
+	if(is_fading)
+		STOP_PROCESSING(SSfastprocess, src)
+	else
+		STOP_PROCESSING(SSprocessing, src)
+	if(target)
+		target.active_music -= src
+		target.update_playing_music()
+		target = null
+
+
+/datum/music/proc/unmask()
+	if(!target || target.playing_music == src)
+		return
+	if(target.playing_music)
+		target.playing_music.mask()
+	target.playing_music = src
+	if(sound_datum || !target || !sound_file)
+		return
+	sound_datum = sound(sound_file, does_loop, 0, CHANNEL_AMBIENT_MUSIC, base_volume * fade_volume)
+	SEND_SOUND(target, sound_datum)
+
+/datum/music/proc/mask()
+	if(target)
+		target.playing_music = null
+		if(sound_datum)
+			SEND_SOUND(target, sound(null, repeat = 0, wait = 0, channel = CHANNEL_AMBIENT_MUSIC))
+			sound_datum = null
+	if(!does_loop)
+		qdel(src)
```

Comment:
Why call qdel() in mask() and then call mask() in destroy()? Either delete them through qdel or delete them through mask and then call the other appropriately (Preferably just call mask in qdel and don't qdel in mask).